### PR TITLE
feat(interpreter)!: make ProtelisProgram immutable

### PR DESCRIPTION
**Pull request by DanySK**
This commit makes Protelis immutable to the `ProtelisProgram` entry point. Mutability is now confined to `ProtelisVM` and `ExecutionContext`. In other words, while `ProtelisVM`s must be handled with care and not shared across structures and threads, entire Protelis programs (namely, abstract syntax trees and module names) can.

### CODE REVIEW

**Comment by DanySK** [comment]
@jakebeal let me know if this change can be merged (rebased). If so, depending whether the CI finished successfully, you can either enable auto-merge (rebase) or rebase it yourself. Otherwise I'll do it.

Thanks.

File: protelis-interpreter/src/main/java/org/protelis/lang/interpreter/impl/AbstractPersistedTree.java [hunk 1]
```diff
@@ -23,8 +23,7 @@
  */
 public abstract class AbstractPersistedTree<S, T> extends AbstractProtelisAST<T> {
 
-    private static final long serialVersionUID = 457607604000217166L;
-    private transient S superscript;
+    private static final long serialVersionUID = 1;
```
Comment (jakebeal) [inline protelis-interpreter/src/main/java/org/protelis/lang/interpreter/impl/AbstractPersistedTree.java:26]: Why this change? Seems unrelated to the rest of the pull.

**Review by jakebeal — APPROVED** [review]

File: protelis-interpreter/src/main/java/org/protelis/lang/interpreter/impl/AbstractPersistedTree.java (hunk 1 above)
Comment (DanySK) [reply protelis-interpreter/src/main/java/org/protelis/lang/interpreter/impl/AbstractPersistedTree.java:26]: I removed a field, potentially changing the serialized format. As such, I modified the serial version UID. This would make systems running Protelis at an older version not compatible with systems using Protelis at this version (cross-version runtime compatibility) if they use the default Java serialization, falling clearly.
